scripts/dss_v2_apdapter.py

```diff
@@ -0,0 +1,609 @@
+
+    def _parse_staging_area(self) -> Tuple[str, str]:
+        """
+        Validate and parse the given staging area URL into bucket and path values.
+        Path value will not have a prefix '/' and will have a postfix '/' if not empty.
+        """
+        split_url = parse.urlsplit(self.args.staging_area)
+        if split_url.scheme != 'gs' or not split_url.netloc:
+            raise ValueError('Staging area must be in gs://<bucket>[/<path>] format')
+        elif split_url.path.endswith('/'):
+            raise ValueError('Staging area URL must not end with a "/"')
+        else:
+            path = f"{split_url.path.lstrip('/')}/" if split_url.path else ''
+            return split_url.netloc, path
+
+    def _get_bucket(self, bucket_name: str) -> gcs.Bucket:
+        """
+        Returns a reference to a GCS bucket.
+        """
+        log.info('Getting bucket: %s', bucket_name)
+        bucket = self.gcs.get_bucket(bucket_name)
+        return bucket
+
+    def _run(self):
+        """
+        Request a list of all bundles in the DSS and process each bundle.
+        """
+        self.skipped_bundles.clear()
+        self.errors.clear()
+
+        prefixes = self.get_prefix_list(self.args.bundle_uuid_prefix, self.args.bundle_uuid_start_prefix)
+
+        params = {
+            'replica': self.dss_src_replica,
+            'per_page': 500,  # limits how many bundles per request min=10 max=500
+        }
+        if prefixes:
+            params['prefix'] = prefixes.pop(0)
+        url_base = self.args.dss_endpoint + '/bundles/all'
+        url = url_base + '?' + parse.urlencode(params)
+
+        future_to_bundle: MutableMapping[Future, BundleFQID] = {}
+        with ThreadPoolExecutor(max_workers=self.args.num_workers) as tpe:
+            while True:
+                log.info('Requesting list of bundles: %s', url)
+                response = requests.get(url)
+                response.raise_for_status()
+                response_json = response.json()
+                for bundle in response_json['bundles']:
+                    bundle_fqid = BundleFQID(uuid=bundle['uuid'], version=bundle['version'])
+                    future = tpe.submit(self._process_bundle, bundle_fqid)
+                    future_to_bundle[future] = bundle_fqid
+                if response_json.get('has_more'):
+                    url = response_json.get('link')
+                elif prefixes:
+                    params['prefix'] = prefixes.pop(0)
+                    url = url_base + '?' + parse.urlencode(params)
+                else:
+                    break
+            for future in as_completed(future_to_bundle):
+                bundle_fqid = future_to_bundle[future]
+                try:
+                    future.result()
+                except CancelledError:
+                    pass  # ignore exception raised when a future is cancelled
+                except BaseException as e:
+                    self.errors[bundle_fqid] = e
```

Comment:
As discussed in the call today, we should add a comment here explaining what the reasoning is for bailing out on the first exception, and why, despite eagerly cancelling all futures, there could be multiple entries in `self.errors`.